### DISCUSSION TITLE
API comparison to Ritual?

### DISCUSSION THREAD

**Discussion by azzamsa** [KDAB/cxx-qt#433] [Q&A] [answered]
Hi.

I tried many example apps from [ritual example repo](https://github.com/rust-qt/examples/tree/master/qml/hello_qml). The code seems very simple compared to the example app in [cxx-qt example](https://github.com/KDAB/cxx-qt/tree/v0.4.1/examples/qml_minimal). As a non-C++ programmer myself, I feel intimidated by the code complexity I find in cxx-qt, I find ritual has simpler code to build a similar app.

- https://github.com/KDAB/cxx-qt/tree/v0.4.1/examples/qml_minimal
- https://github.com/rust-qt/examples/tree/master/qml/hello_qml

Is this because of the full potential described in the `introduction` chapter of the book https://kdab.github.io/cxx-qt/book/index.html? Or because it uses `safe rust` compared to ritual as described in this comparison https://github.com/KDAB/cxx-qt#comparison-to-other-rust-qt-bindings?

Are there any more points I need to know about why the CXX-Qt code looks so complex compared to ritual?

Thanks for maintaining CXX-Qt :heart:

**Comment by LeonMatthesKDAB** [comment]
The comparison you have right there is a bit of an apples-to-oranges comparison.
The ritual app you posted doesn't actually have any Rust behavior, apart from spinning up the QML app, whilst the CXX-Qt version does:
- Integration with CMake build system.
- Define a QObject subclass in Rust
- Expose the QObject subclass to QML
- Actually use the QObject subclass in QML
- Adds some tests

A fairer comparison would be between the ritual example and the [`cargo_without_cmake` example from @Be-ing 's newest PR](https://github.com/Be-ing/cxx-qt/tree/qml_element/examples/cargo_without_cmake).

Furthermore, you could get rid of the CMakeLists.txt file in that example (it's just for integrating into the overall cxx-qt repo).
Also, the `src/cxxqt_object.rs` does something that the given ritual example doesn't do, so you could ignore that.

Then you basically end up with the same code your ritual example had :)
Relevant files being only:
- build.rs
- Cargo.toml
- src/main.rs
- and the `qml/` folder.

**Comment by Be-ing** [comment, marked as answer]
With the new 0.5.0 release, it is now possible to write QML applications in Rust without needing C++ boilerplate. Take a look at the [cargo_without_cmake example](https://github.com/KDAB/cxx-qt/tree/main/examples/cargo_without_cmake).

**Reply by azzamsa** [reply]
The new v0.5.0 release reduce many boilerplate code. Awesome!
